add_keys.py

```diff
@@ -0,0 +1,204 @@
+
+
+def init_account(w3: Web3):
+    '''
+    Initialise the signer.
+    '''
+    privkey = os.getenv('PRIV_KEY')
+    account = w3.eth.account.from_key(privkey)
+    logging.info(f'Account: {account.address}')
+    return account
+
+
+
+@click.command()
+@click.option("--nonce", default=-1, type=int, help="Nonce to use (default uses next nonce on chain)")
+@click.option("--gas", default=GAS_LIMIT, type=int, help="Gas limit - defaults to {}, which is suitable for submitting 100 keys.".format(GAS_LIMIT))
+@click.option("--operator", help="Operator ID.", type=int)
+@click.option("--eth1-uri", help="Ethereum node address", type=str)
+@click.option("--pool-address", help="Lido pool contract address - defaults to mainnet contract address", type=str, default=POOL_ADDRESS)
+@click.argument('filename')
+def main(filename, nonce, operator, gas, eth1_uri, pool_address):
+    '''
+    Main entrypoint.
+    '''
+    try:
+        if eth1_uri == None:
+            eth1_uri = os.environ.get('WEB3_PROVIDER_URI', None)
+            if eth1_uri == None:
+                raise Exception("Must pass --eth1-uri flag or set WEB3_PROVIDER_URI environment variable")
+        w3 = init_provider(eth1_uri)
+        registry = init_contracts(w3, pool_address)
+        account = init_account(w3)
+
+        if operator is None:
+            raise Exception("--operator flag is required")
+
+        with open(filename, 'r') as f:
+            keyfile = json.load(f)
+            f.close()
+
+        keys = [x.get('pubkey') for x in keyfile]
+        signatures = [x.get('signature') for x in keyfile]
+        logging.info("Found %d keys...", len(keys))
+
+        tx = build_tx(registry, operator, keys, signatures, account, gas)
+        send_tx(w3, tx, account, nonce)
+    except Exception as e:
+        print("Error: {}".format(e))
+        sys.exit(1)
+
+
+def build_tx(registry, operator_id, keys, signatures, account, gas):
+    '''
+    Build an addSigningKeysOperatorBH tx from the keys and signatures provided.
+    '''
+
+    # sanity_check
+    if len(keys) != len(signatures):
+        raise Exception("Imbalance in keys and signatures")
+
+    return registry.functions.addSigningKeysOperatorBH(
+        operator_id,
+        len(keys),
+        "".join(keys),
+        "".join(signatures)
+    ).buildTransaction({
+        'from': account.address,
+        'gas': gas
+    })
+
+def send_tx(w3, tx, account, nonce):
+    '''
+    Send the contract interaction.
+    '''
+    try:
+        # execute tx locally to check validity
+        w3.eth.call(tx)
+        logging.info('Calling tx locally succeeded.')
+        print(f'Tx data: {tx.__repr__()}')
+        if prompt('Should we send this TX? [y/n]: ', ''):
+            sign_and_send_tx(w3, tx, account, nonce)
+    except SolidityError as sl:
+        str_sl = str(sl)
+        logging.error(f'Calling tx locally failed: {str_sl}')
+    except ValueError as exc:
+        (args, ) = exc.args
+        if args["code"] == -32000:
```

Comment:
What is this error -32000? And why does it need special handling if we bail out anyway?

Reply:
code 32000 errors are coming specifically from geth, so we want the exception to pass through; other errors we just report as being unexpected errors. This code mostly came from the existing lido contract interactions from the oracle.